src/consensus/aft/raft.h

```diff
@@ -284,6 +282,32 @@ namespace aft
       return replica_state == kv::ReplicaState::Follower;
     }
 
+    bool is_eligible_voter(const ccf::NodeId& from) const
+    {
+      if (from == state->my_node_id)
+      {
+        return !catching_up && !retiring;
+      }
+      else
+      {
+        auto it = nodes.find(from);
+        return it != nodes.end() && !it->second.catching_up;
+      }
+    }
+
+    size_t num_eligigble_voters() const
+    {
+      size_t r = (catching_up || retiring) ? 0 : 1;
+      for (const auto& [_, state] : nodes)
+      {
+        if (!state.catching_up)
+        {
+          r++;
+        }
+      }
+      return r;
+    }
```

Comment:
Is there a reason this duplicates some of `is_eligible_voter`? While less efficient, I think this would be much more readable as

```suggestion
    size_t num_eligigble_voters() const
    {
      size_t r = 0;
      for (const auto& id, _] : nodes)
      {
        if (is_eligible_voter(id))
        {
          r++;
        }
      }
      return r;
    }
```

This could also be replaced by `return std::count_if(...)`.